File: server/downloaders/synologyDownloadStation/synologyDownloadStation.go
package synologyDownloadStation

import (
	"encoding/json"
	"errors"
	"io/ioutil"
	"net/http"
	"net/url"
	"strings"
	"holerr/core/config"
	"holerr/core/db"
	"holerr/core/log"
	"holerr/downloaders/downloader"

	"github.com/monaco-io/request"
)

var cfg = config.Get()
var synoCfg = cfg.Downloaders.SynologyDownloadStation
var ENDPOINT = synoCfg.Endpoint

type SynologyDownloadStation struct {
	downloader.Downloader
}

func New() SynologyDownloadStation {
	s := SynologyDownloadStation{}
	sid, err := connect("DownloadStation")
	if err != nil {
		log.Fatal(err)
	}
	log.Info("DownloadStation session ID: " + sid)
	return s
}

func (r SynologyDownloadStation) IsConnected() bool {
	sid, err := connect("DownloadStation")
	return err == nil && sid != ""
}

func (s SynologyDownloadStation) AddDownload(uri string, name string, preset config.Preset) (string, error) {
	sid, err := connect("DownloadStation")
	if err != nil {
		log.Fatal(err)
	}

	req, err := http.NewRequest(http.MethodGet, getApiUrl("/DownloadStation/task.cgi"), nil)
	if err != nil {
		return "", err
	}

	q := req.URL.Query()
	q.Add("api", "SYNO.DownloadStation.Task")
	q.Add("version", "1")
	q.Add("method", "create")
	q.Add("uri", uri)
	q.Add("_sid", sid)

	destination := preset.OutputDir
	if destination != "" {
		if preset.CreateSubDir {
			err := createOutputDir(destination, name)
			if err != nil {
				return "", err
			}
			destination = destination + "/" + name
		}
		q.Add("destination", destination)
	}
	req.URL.RawQuery = q.Encode()

	statusCode, body, err := makeRequest(req)
	if err != nil {
		return "", err
	}
	if statusCode != 200 {
		return "", errors.New("Could not add download")
	}
	var obj Status
	jsonError := json.Unmarshal(body, &obj)
	if jsonError != nil {
		return "", jsonError
	}
	if !obj.Success {
		return "", errors.New("Error while adding download")
	}

	return getId(uri)
}

func (s SynologyDownloadStation) GetTaskStatus(id string) (int, int, error) {
	sid, err := connect("DownloadStation")
	if err != nil {
		log.Fatal(err)
	}
	client := request.Client{
		URL: "/DownloadStation/task.cgi",
		Params: map[string]string{
			"api":        "SYNO.DownloadStation.Task",
			"version":    "1",
			"method":     "getinfo",
			"additional": "transfer",
			"id":         id,
			"_sid":       sid,
		},
	}
	prepareClient(&client)
	resp, err := client.Do()
	if err != nil {
		return 0, 0, err
	}
	if resp.Code != 200 {
		return 0, 0, errors.New("Could retrieve download info")
	}

	var obj Tasks
	jsonError := json.Unmarshal(resp.Data, &obj)
	if jsonError != nil {
		return 0, 0, jsonError
	}
	if !obj.Success {
		return 0, 0, errors.New("Error while get download info")
	}

	downloaderStatusMapping := map[string]int{
		downloader.DownloadStatus["WAITING"]:       db.DownloadStatus["DOWNLOADER_DOWNLOADING"],
		downloader.DownloadStatus["DOWNLOADING"]:   db.DownloadStatus["DOWNLOADER_DOWNLOADING"],
		downloader.DownloadStatus["PAUSED"]:        db.DownloadStatus["DOWNLOADER_DOWNLOADING"],
		downloader.DownloadStatus["FINISHING"]:     db.DownloadStatus["DOWNLOADER_DOWNLOADING"],
		downloader.DownloadStatus["FINISHED"]:      db.DownloadStatus["DOWNLOADER_DOWNLOADED"],
		downloader.DownloadStatus["HASH_CHECKING"]: db.DownloadStatus["DOWNLOADER_DOWNLOADING"],
		downloader.DownloadStatus["SEEDING"]:       db.DownloadStatus["DOWNLOADER_DOWNLOADED"],
		downloader.DownloadStatus["EXTRACTING"]:    db.DownloadStatus["DOWNLOADER_DOWNLOADING"],
		downloader.DownloadStatus["ERROR"]:         db.DownloadStatus["ERROR_DOWNLOADER"],
	}
	return downloaderStatusMapping[obj.Data.Tasks[0].Status], obj.Data.Tasks[0].Additional.Transfer.SizeDownloaded, nil
}

func (s SynologyDownloadStation) DeleteDownload(id string) error {
	sid, err := connect("DownloadStation")
	if err != nil {
		log.Fatal(err)
	}

	client := request.Client{
		URL: "/DownloadStation/task.cgi",
		Params: map[string]string{
			"api":     "SYNO.DownloadStation.Task",
			"version": "1",
			"method":  "delete",
			"id":      id,
			"_sid":    sid,
		},
	}
	prepareClient(&client)
	resp, err := client.Do()
	if err != nil {
		return err
	}
	if resp.Code != 200 {
		return errors.New("Could not delete download task")
	}

	return nil
}

func prepareClient(client *request.Client) {
	client.URL = ENDPOINT + "/webapi" + client.URL

	if client.Method == "" {
		client.Method = "GET"
	}
}

func getApiUrl(path string) string {
	return ENDPOINT + "/webapi" + path
}

func connect(session string) (string, error) {
	req, err := http.NewRequest(http.MethodGet, getApiUrl("/auth.cgi"), nil)
	if err != nil {
		return "", err
	}

	q := url.Values{}
	q.Add("api", "SYNO.API.Auth")
	q.Add("version", "3")
	q.Add("method", "login")
	q.Add("session", session)
	q.Add("account", synoCfg.Username)
	q.Add("passwd", synoCfg.Password)
	q.Add("format", "sid")
	req.URL.RawQuery = q.Encode()

	statusCode, body, err := makeRequest(req)
	if err != nil {
		return "", err
	}

	if statusCode != 200 {
		return "", errors.New("Could not login to " + session)
	}

	var obj Auth
	jsonError := json.Unmarshal(body, &obj)
	if jsonError != nil {
		return "", jsonError
	}
	if !obj.Success {
		return "", errors.New("Error while login to " + session)
	}
	return obj.Data.Sid, nil
}

func makeRequest(req *http.Request) (int, []byte, error) {
	req.URL.RawQuery = strings.ReplaceAll(req.URL.RawQuery, "+", "%20")

	client := &http.Client{}
	resp, err := client.Do(req)
	if err != nil {
		return 0, []byte{}, err
	}
	defer resp.Body.Close()

	body, err := ioutil.ReadAll(resp.Body)
	return resp.StatusCode, body, err
}

func getId(uri string) (string, error) {
	sid, err := connect("DownloadStation")
	if err != nil {
		log.Fatal(err)
	}

	client := request.Client{
		URL: "/DownloadStation/task.cgi",
		Params: map[string]string{
			"api":        "SYNO.DownloadStation.Task",
			"version":    "1",
			"method":     "list",
			"additional": "detail",
			"_sid":       sid,
		},
	}
	prepareClient(&client)
	resp, err := client.Do()
	if err != nil {
		return "", err
	}
	if resp.Code != 200 {
		return "", errors.New("Could not list downloads")
	}
	var obj Tasks
	jsonError := json.Unmarshal(resp.Data, &obj)
	if jsonError != nil {
		return "", jsonError
	}
	if !obj.Success {
		return "", errors.New("Error while search download id")
	}

	for _, task := range obj.Data.Tasks {
		if task.Additional.Detail.Uri == uri {
			return task.Id, nil
		}
	}

	return "", errors.New("Download not found")
}


func createOutputDir(parent string, name string) error {
	sid, err := connect("FileStation")
	if err != nil {
		log.Fatal(err)
	}

	req, err := http.NewRequest(http.MethodGet, getApiUrl("/entry.cgi"), nil)
	if err != nil {
		return err
	}

	folder_path := parent
	if(folder_path[0] != '/') {
		folder_path = "/" + folder_path
	}

	q := req.URL.Query()
	q.Add("api", "SYNO.FileStation.CreateFolder")
	q.Add("version", "2")
	q.Add("method", "create")
	q.Add("folder_path", folder_path)
	q.Add("name", name)
	q.Add("_sid", sid)
	req.URL.RawQuery = q.Encode()

	statusCode, body, err := makeRequest(req)
	if err != nil {
		return err
	}
	if statusCode != 200 {
		return errors.New("Could not create sub folder")
	}
	var obj Status
	jsonError := json.Unmarshal(body, &obj)
	if jsonError != nil {
		return jsonError
	}
	if !obj.Success {
		return errors.New("Error while creating sub folder")
	}

	return nil
}

File: server/downloaders/synology_download_station.py
from .downloader import Downloader
from server.core import config
from server.core.log import Log
from .synology_download_station_models import Auth, Tasks, Status
from server.core.config_models import Preset


import requests
from pathlib import Path
import urllib

log = Log.get_logger(__name__)


class SynologyDownloadStation(Downloader):
    def __init__(self, config):
        pass

    def get_name(self) -> str:
        return "Synology Download Station"

    def is_connected(self) -> bool:
        sid = self._connect("DownloadStation")
        return sid is not None

    def add_download(self, uri: str, title: str, preset: Preset) -> str:
        sid = self._connect("DownloadStation")
        if sid is None:
            raise Exception("Could not connect to Synology Download Station")

        params = {
            "api": "SYNO.DownloadStation.Task",
            "version": "1",
            "method": "create",
            "uri": uri,
            "_sid": sid,
        }
        if preset.output_dir is not None:
            destination = preset.output_dir
            if preset.create_sub_dir:
                sub_folder = self.get_sub_folder_name(title)
                self._create_output_dir(destination, sub_folder)
                destination += "/" + sub_folder
            params["destination"] = destination

        # known bug in Synology Download Station API fails with "+" destination
        params = urllib.parse.urlencode(params).replace("+", "%20")
        res = self._call("/DownloadStation/task.cgi", params=params)
        if res.status_code != 200:
            raise Exception("Could not add download " + res)
        obj = Status(**res.json())
        if not obj.success:
            log.debug(res.request.url)
            raise Exception("Error while adding download, code: " + str(obj.error.code))
        return self._get_download_id(uri)

    def get_task_status(self, id: str) -> tuple[str, int]:
        sid = self._connect("DownloadStation")
        if sid is None:
            raise Exception("Could not connect to Synology Download Station")

        params = {
            "api": "SYNO.DownloadStation.Task",
            "version": "1",
            "method": "getinfo",
            "additional": "transfer",
            "id": id,
            "_sid": sid,
        }
        res = self._call("/DownloadStation/task.cgi", params=params)
        if res.status_code != 200:
            raise Exception("Could not get task status " + str(res.status_code))

        obj = Tasks(**res.json())
        if not obj.success:
            raise Exception(
                "Error while getting task status, code: " + str(obj.error.code)
            )

        task = obj.data.tasks[0]

        return task.status, task.additional.transfer.size_downloaded

    def delete_download(self, id: str):
        sid = self._connect("DownloadStation")
        if sid is None:
            raise Exception("Could not connect to Synology Download Station")

        params = {
            "api": "SYNO.DownloadStation.Task",
            "version": "1",
            "method": "delete",
            "id": id,
            "_sid": sid,
        }
        res = self._call("/DownloadStation/task.cgi", params=params)
        if res.status_code != 200:
            raise Exception(
                "Could not delete download " + id + " " + str(res.status_code)
            )

    def _call(self, path, **kwargs):
        return requests.request("GET", self._get_api_url(path), **kwargs)

    def _get_api_url(self, path: str) -> str:
        return config.downloader.synology_download_station.endpoint + "/webapi" + path

    def _connect(self, session: str) -> str | None:
        syno_cfg = config.downloader.synology_download_station
        params = {
            "api": "SYNO.API.Auth",
            "version": "3",
            "session": session,
            "account": syno_cfg.username,
            "passwd": syno_cfg.password,
            "format": "sid",
            "method": "login",
        }
        res = self._call("/auth.cgi", params=params)
        if res.status_code != 200:
            raise Exception(
                "Could not login to " + session + " " + str(res.status_code)
            )

        auth = Auth(**res.json())
        return auth.data.sid or None

    def _get_download_id(self, uri: str) -> str:
        sid = self._connect("DownloadStation")
        if sid is None:
            raise Exception("Could not connect to Synology Download Station")

        params = {
            "api": "SYNO.DownloadStation.Task",
            "version": "1",
            "format": "sid",
            "method": "list",
            "additional": "detail",
            "_sid": sid,
        }

        res = self._call("/DownloadStation/task.cgi", params=params)
        if res.status_code != 200:
            raise Exception("Could not list downloads, " + str(res.status_code))
        obj = Tasks(**res.json())
        if not obj.success:
            raise Exception(
                "Error while searching download id, code: " + str(obj.error.code)
            )

        for task in obj.data.tasks:
            if task.additional.detail.uri == uri:
                return task.id
        log.debug(f"Download {uri} not found")
        return None

    def _create_output_dir(self, parent: str, name: str):
        sid = self._connect("FileStation")
        if sid is None:
            raise Exception("Could not connect to Synology File Station")

        folder_path = parent
        if folder_path[0] != "/":
            folder_path = "/" + folder_path

        params = {
            "api": "SYNO.FileStation.CreateFolder",
            "version": "2",
            "method": "create",
            "folder_path": folder_path,
            "name": name,
            "_sid": sid,
        }
        res = self._call("/entry.cgi", params=params)
        if res.status_code != 200:
            raise Exception("Could not create folder " + str(res.status_code))
        obj = Status(**res.json())
        if not obj.success and obj.error.code != 109:  # 109 = folder already exists
            log.debug(res.request.url)
            raise Exception("Error while creating folder, code: " + str(obj.error.code))

    @staticmethod
    def get_sub_folder_name(name: str) -> str:
        return Path(name).stem
